File: algorithms/medium/smallest_number_in_infinite_set.ex
defmodule SmallestInfiniteSet do
  use GenServer

  @spec init_() :: any
  def init_() do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
    GenServer.call(__MODULE__, :init)
  end

  @spec pop_smallest() :: integer
  def pop_smallest() do
    GenServer.call(__MODULE__, :pop)
  end

  @spec add_back(num :: integer) :: any
  def add_back(num) do
    GenServer.call(__MODULE__, {:add, num})
  end

  def init(state) do
    {:ok, state}
  end

  def handle_call(:init, _from, _state) do
    {:reply, true, {:gb_sets.empty(), 1}}
  end

  def handle_call(:pop, _from, {set, curr}) do
    if :gb_sets.is_empty(set) do
      {:reply, curr, {set, curr + 1}}
    else
      {r, set} = :gb_sets.take_smallest(set)
      {:reply, r, {set, curr}}
    end
  end

  def handle_call({:add, num}, _from, {set, curr}) do
    if num < curr && !:gb_sets.is_member(num, set) do
      {:reply, num, {:gb_sets.add(num, set), curr}}
    else
      {:reply, num, {set, curr}}
    end
  end
end

# Your functions will be called as such:
# SmallestInfiniteSet.init_()
# param_1 = SmallestInfiniteSet.pop_smallest()
# SmallestInfiniteSet.add_back(num)

# SmallestInfiniteSet.init_ will be called before every test case, in which you can do some necessary initializations.

File: algorithms/medium/smallest_number_in_infinite_set.rs
use std::collections::{HashSet, BinaryHeap};
use std::cmp::Reverse;

#[derive(Default)]
struct SmallestInfiniteSet {
    smallest: i32,
    set: HashSet<i32>,
    pq: BinaryHeap<Reverse<i32>>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl SmallestInfiniteSet {

    fn new() -> Self {
        Self {
            smallest: 1,
            ..Default::default()
        }
    }
    
    fn pop_smallest(&mut self) -> i32 {        
        if !self.pq.is_empty() {
            let n = self.pq.pop().unwrap().0;
            self.set.remove(&n);
            return n;
        }

        self.smallest += 1;
        
        self.smallest - 1
    }
    
    fn add_back(&mut self, num: i32) {
        if num < self.smallest && !self.set.contains(&num) {
            self.pq.push(Reverse(num));
            self.set.insert(num);
        }        
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * let obj = SmallestInfiniteSet::new();
 * let ret_1: i32 = obj.pop_smallest();
 * obj.add_back(num);
 */
